### PR TITLE
Performance "Seamless loading of gzipped JSON/txt files (data and map)"

### PR DESCRIPTION
**Pull request by int-ua**
#### Summary

Performance "Loading of gzipped world/maps/x.y.z/x.y.z.map and data/json files"

#### Purpose of change

Ability to drastically reduce save folder sizes with (hopefully) only a marginal performance impact.

#### Describe the solution

Checking first two bytes of files before loading JSON and if they are compressed use `gzstream` instead of `std::ifstream` to read them.
Reference: https://github.com/kevingranade/Cataclysm/commit/e751f0bc4acd52cbbf6e77dc2610d2aa98b232ef
Supersedes #43986 with all the commits squashed and proper `Co-authored-by`s.

Optional compression by the game during save action will be added in a future PR or here if it stays open long enough.

#### Testing

```sh
#!/bin/bash
WORLD_NAME="test"

# verify that world loads without issues
./cataclysm-tiles --world "$WORLD_NAME"

# create a copy of the world and gzip all maps/ files in the copy
cp -r "save/${WORLD_NAME}" "save/${WORLD_NAME}_gz"
find "save/${WORLD_NAME}_gz/maps/" -type f -exec gzip "{}" \; -exec mv "{}.gz" "{}" \;

# try loading the gzipped copy
./cataclysm-tiles --world "${WORLD_NAME}_gz"

# compare sizes
du -sh "save/${WORLD_NAME}/maps/"
du -sh "save/${WORLD_NAME}_gz/maps/"
```

### CODE REVIEW

**Comment by int-ua** [comment]
Wait, looks like it no longer loads the world for some reason. :grey_question:

File: src/cata_utility.cpp [hunk 1]
```diff
@@ -346,7 +348,20 @@ bool read_from_file( const std::string &path, const std::function<void( std::ist
...
-        reader( fin );
+
+        // check if file is gzipped
+        // (byte1 == 0x1f) && (byte2 == 0x8b)
+        char header[2];
+        fin.read( header, 2 );
+        if( ( header[0] == '\x1f' ) && ( header[1] == '\x8b' ) ) {
+            const char *cstr_path = path.c_str();
+            igzstream fingz( cstr_path, std::ios::binary );
+            reader( fingz );
```
Comment (olanti-p) [inline src/cata_utility.cpp:359]: `igzstream` always opens in binary mode (see `gzstreambuf::open`). You're passing some value here, and that overwrites default one that specifies the 'in' mode (see `igzstream` constructor), breaking the reading. Seems kinda weird to me to make it a default argument, but whatever

Curiously, fixing that wasn't enough. The game managed to load 2 compressed .map files, but got stuck on the 3rd one, entering some kind of infinite loop somewhere inside json error handling code (deep enough to confuse gdb). 

Changing the code to unpack the whole file into memory, and then read json from memory:
```suggestion
            igzstream fingz( path.c_str() );
            std::stringstream ss;
            ss << fingz.rdbuf();
            if( fingz.bad() ) {
                throw std::runtime_error( "reading file failed" );
            }
            ss.seekg( 0, std::ios::beg );
            reader( ss );
```
seems to be working, and the game loads compressed save successfully. Not sure, why, though. Maybe the reader uses some kind of seeking, and that doesn't work in igzstream?

Comment (int-ua) [reply src/cata_utility.cpp:359]: I was digging with `gdb` too and with `igzstream fingz( cstr_path );` `jsin->tell()` at `json.cpp:83` was returning `ffffffff`. Anyway, I'll try your code now, thanks again.

**Comment by int-ua** [comment]
Ok, looks like it finally worked. But if we unpack it into memory anyway does that mean that `gzstream` is absolutely useless here and can be just replaced with `zlib` functions?

**Comment by ZhilkinSerg** [comment]
It works, but I would like to see optional compressed file saving code implemented here too (we should add tests for both save methods later) plus getting rid of gzstream dependency and using zlib directly would be very nice.

**Comment by int-ua** [comment]
Can this work here? https://github.com/mapbox/gzip-hpp (don't have much time right now, will get back to it later)

**Comment by actual-nh** [comment]
As someone low on disk space, any progress on this?

**Comment by int-ua** [comment]
@Aivean The main thing missing here is the compression on saving.

Less pressing issue is the fact that we had to load whole files into memory because `gzstream` was not working properly for **some** reason and now that whole idea of having `gzstream` is kind of moot. Can be solved either by making `gzstream` work properly without loading whole files into memory or by replacing `gzstream` with something else.

Maybe it's not so trivial after all, no pressure.

**Comment by Aivean** [comment]
Thanks for the explanation. I'll try to understand why `gzstream` is not working. However, no promises as to when.

**Comment by jbytheway** [comment]
We generally prefer to have fewer dependencies, and if we're not actually streaming the input then I agree with @ZhilkinSerg that it's probably better to just use zlib directly rather than make gzstream work.  I'm guessing that the code required to use zlib to decompress an entire file into a `std::string` should be pretty simple and be a lot less than folding `gzstream` into the project.

On the other hand, if `gzstream` would make the output of compressed files substantially easier then perhaps it's still worth having.

**Comment by ZhilkinSerg** [comment]
Let's move it to post-0.F

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -262,6 +263,7 @@ if (MINGW AND NOT RELEASE)
                     ${RuntimeLib_mpeg})
         endif ()
     endif ()
+
```
Comment (int-ua) [inline src/CMakeLists.txt:266]:
```suggestion
```

**Comment by int-ua** [comment]
I'm trying to debug it in QtCreator and I've just confirmed that this change would also allow compressing all text files, not just maps.

**Comment by int-ua** [comment]
I've reverted https://github.com/CleverRaven/Cataclysm-DDA/pull/44218/commits/b16f2f72e393caa91e3317125d874a95a24ce3d6 and trying to debug it. At that state the buffer is empty on `fingz` when it is already being read for some reason.

After a couple of changes of questionable value I assume that `JsonIn.seek` or something else is trying to rewind the `igzstream` and fails.
```cpp
        if( ( header[0] == '\x1f' ) && ( header[1] == '\x8b' ) ) {
            igzstream fingz( path.c_str() );
            reader( fingz );
        } else {
```

**Comment by int-ua** [comment]
This convoluted mess

```cpp
char header[2];
        fin.read( header, 2 );
        fin.clear();
        fin.seekg( 0, std::ios::beg ); // reset read position

        if( ( header[0] == '\x1f' ) && ( header[1] == '\x8b' ) ) {
            std::ostringstream deflated_contents_stream;
            std::string str;

            deflated_contents_stream << fin.rdbuf();
            str = deflated_contents_stream.str();

            z_stream zs;
            memset(&zs, 0, sizeof(zs));

            if (inflateInit(&zs) != Z_OK)
                throw(std::runtime_error("inflateInit failed while decompressing."));

            zs.next_in = reinterpret_cast<unsigned char*>(const_cast<char*>(str.data()));
            zs.avail_in = str.size();

            int ret;
            char outbuffer[32768];
            std::string outstring;

            // get the decompressed bytes blockwise using repeated calls to inflate
            do {
                zs.next_out = reinterpret_cast<Bytef*>(outbuffer);
                zs.avail_out = sizeof(outbuffer);

                ret = inflate(&zs, 0);

                if (outstring.size() < zs.total_out) {
                    outstring.append(outbuffer,
                                     zs.total_out - outstring.size());
                }

            } while (ret == Z_OK);

            inflateEnd(&zs);

            if (ret != Z_STREAM_END) { // an error occurred that was not EOF
                std::ostringstream oss;
                oss << "Exception during zlib decompression: (" << ret << ") "
                    << zs.msg;
                throw(std::runtime_error(oss.str()));
            }
            
            std::stringstream inflated_contents_stream;
            inflated_contents_stream << outstring;

            reader( inflated_contents_stream );
```
results in zs.msg being "Incorrect header check" and I have no idea why yet.

**Comment by int-ua** [comment]
There is now also a PR that will add an option to use a more compact binary serialization instead of JSON: #50143

The author, @akrieger, said that both approaches can be combined. I'll wait until he has time to review the code in this PR.

**Comment by ZhilkinSerg** [comment]
This will come in handy with what's following after #51771

**Review by akrieger — COMMENTED** [review]
I had a long comment then GitHub ate it while I was clicking between changes and text. So you get a short comment instead.

Implementation looks sound. Please add `zlib` to `msvc-full-features/vcpkg.json` and `msvc-object_creator/vcpkg.json`, because although `zlib` is ubiquitous on other platforms it's being indirectly provided for the Windows builds. I don't know what the check failures are about. I'm unhappy with the performance because there's a ton of string copies happening and a lack of ability to reserve or efficiently move things (moves in/out of stringstream are available with c++20). Ultimately though seems like the tradeoff is some load time regression in exchange for substantial disk size gains, which is especially important for save data and maybe less so for static data.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -205,14 +205,17 @@ if (MINGW AND NOT RELEASE)
...
     find_library(RuntimeLib_GCC_S_DW2_1  "gcc_s_dw2-1")
     find_library(RuntimeLib_STDC_PP_6    "stdc++-6")
     find_library(RuntimeLib_WINPTHREAD_1 "winpthread-1")
+    find_library(RuntimeLib_zlib  "zlib1")
+
     set(RuntimeLib_GCC_ALL
             ${RuntimeLib_GCC_S_DW2_1}
             ${RuntimeLib_STDC_PP_6}
-            ${RuntimeLib_WINPTHREAD_1})
+            ${RuntimeLib_WINPTHREAD_1
```
Comment (ZhilkinSerg) [inline src/CMakeLists.txt:217]:
```suggestion
            ${RuntimeLib_WINPTHREAD_1}
```

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -205,14 +205,17 @@ if (MINGW AND NOT RELEASE)
...
+            ${RuntimeLib_zlib})}
```
Comment (ZhilkinSerg) [inline src/CMakeLists.txt:218]:
```suggestion
            ${RuntimeLib_WINPTHREAD_1}
            ${RuntimeLib_zlib})
```

**Comment by BrettDong** [comment]
> @BrettDong why did this need to get reverted?

Clicked the wrong button and GitHub automatically generates a "revert" commit references this pull request. No worries, this is never reverted.

**Comment by int-ua** [comment]
So. Compressing files in `save/World/maps` works. Although they get uncompressed when written to you can already reduce size of the old regions until you visit them again if ever. But JSON files in `save/World/` and `data/` directory didn't work 100%. Somehow only a handful of the latter load successfully in the latest experimental. I didn't have time when this PR was just merged. For example `data/json/achievements.json` doesn't load when gzipped and starts with `0x1F8B` with the error "expected object or array". Not sure why yet.

**Pull request merged**